feat(hub): list collections

quanghuynt14 opened the pull request:
Issue: https://github.com/huggingface/huggingface.js/issues/271

This PR only includes the call of list collections. Why?
- It's one of my early contributions to the project, so I think I should split the [PR](https://github.com/huggingface/huggingface.js/pull/367) of @hackpk into many small PRs.

- > I was really hoping to be able to make a collection of our best models and datasets, so that they can be displayed on our landing page.
- I think list collections solve the problem of displaying collections in real time 

I’m having a bit of trouble with the type, and I need to search through the project to understand it better. We don't have any public schema API?

Thanks for review it.

quanghuynt14 commented on packages/hub/src/lib/list-collections.ts:
```diff
@@ -0,0 +1,132 @@
+	gating: boolean;
+	lastUpdated: Date;
+	owner: string;
+	items: {
```
the item type is inspired by the python version: https://github.com/huggingface/huggingface_hub/blob/c8ffea9ca815b10edf07085443d524bac479153d/src/huggingface_hub/hf_api.py#L1189

quanghuynt14 commented on packages/hub/src/lib/list-collections.ts:
```diff
@@ -0,0 +1,136 @@
+
+	const search = new URLSearchParams([
+		...Object.entries({
+			limit: String(Math.min(totalToFetch, 100)),
```
by the [doc](https://huggingface.co/docs/hub/api#get-apicollections), max number of collections par page is 100.

coyotte508 replied:
technically it can be 10000 if all of:

- `owner` is set
- `expand` is set to false explicitly

but not sure we want to bother

quanghuynt14 commented on packages/hub/src/types/api/api-collection.ts:
```diff
@@ -0,0 +1,22 @@
+	gating: boolean;
+	lastUpdated: string;
+	owner: {
+		name: string;
```
The owner type includes more properties, but I only provided the `name` here. Then, format it to the `CollectionEntry` type, where owner is a string representing the name.

coyotte508 reviewed (commented):
Thanks for the PR!

I will look into providing a schema for the existing collection APIs

quanghuynt14 commented:
> I will look into providing a schema for the existing collection APIs

Do we use `json-schema` or something similar? 

We can use `json-schema-to-typescript` togenerate the type.

coyotte508 commented:
Hey @quanghuynt14 , here's an OpenAPI JSON spec for some of the hub endpoints, including collection ones: https://gist.github.com/coyotte508/ec7d12713d94a4fe4988afa3fffc23f1

It includes json schema for the various responses & requests, and you can also load the openapi schema on https://editor-next.swagger.io/ for example

do you think it's helpful for your PRs?

quanghuynt14 commented:
I checked the schema and found a mismatch in the schema of `"/api/collections/{namespace}/{slug}-{id}"`: 

We have:
-  requipred prop `"position": { "type": "number" }` 
-  requipred prop `"shareUrl": { "type": "string" }`

in the `collection` object.

After verifying `"slug": "google/gemma-3n-685065323f5984ef315c93f4"`, I found that it does not include this property.
```json
{
    "slug": "google/gemma-3n-685065323f5984ef315c93f4",
    "title": "Gemma 3n",
    "description": "",
    "gating": false,
    "lastUpdated": "2025-06-26T15:55:44.512Z",
    "owner": {
      ...
    },
    "items": [
      ...
    ],
    "theme": "purple",
    "private": false,
    "upvotes": 146,
    "isUpvotedByUser": false
  },
```

These 2 props (`position` and `shareUrl`) are present only in `items` of type `"collection"`. For example:
```json
{
    "slug": "kristenq/reasoning-684330e0ce0c4e30fe59456a",
    "title": "Reasoning",
    "description": "Advanced reasoning models. ",
    "gating": false,
    "lastUpdated": "2025-06-06T18:18:26.733Z",
    "owner": {
      "_id": "678fdd3a44c9b7e8161b460e",
      "avatarUrl": "/avatars/abad829ad5a3efb420180cd6e3217873.svg",
      "fullname": "Kristen Quan",
      "name": "kristenq",
      "type": "user",
      "isPro": false,
      "isHf": false,
      "isHfAdmin": false,
      "isMod": false
    },
    "items": [
      {
        "_id": "684330f27ce524322498baa7",
        "position": 0,
        "type": "collection",
        "id": "67ee7145ec3d31f7c7a75cab",
        "slug": "Tesslate/synthia-s1-reasoning-model-67ee7145ec3d31f7c7a75cab",
        "title": "Synthia-S1 REASONING MODEL",
        "description": "Creative, Scientific, and Coding",
        "lastUpdated": "2025-04-03T11:31:05.262Z",
        "numberItems": 3,
        "owner": {
          ...
        },
        "theme": "blue",
        "shareUrl": "https://huggingface.co/collections/Tesslate/synthia-s1-reasoning-model-67ee7145ec3d31f7c7a75cab", 
        "upvotes": 3,
        "isUpvotedByUser": false
      }
    ],
    "theme": "green",
    "private": false,
    "upvotes": 0,
    "isUpvotedByUser": false
  },
```

Here is the corrected schema: https://gist.github.com/quanghuynt14/1c55f44978248bc12889b1bde2359c0c

- I deleted prop `position` in the collection object.
- I moved prop `shareUrl` into `items` type `collection`.

quanghuynt14 commented:
You can review the PR. I think the type is now complete.

coyotte508 reviewed (commented):
Thanks a lot!

I think just one change on how `owner` and `item` params are handled and it'll be ready to merge (I'll still re-review)

coyotte508 commented on packages/hub/src/types/api/api-collection.ts:
```diff
@@ -0,0 +1,462 @@
+		}
+	);
+	owner:
+	| {
+		avatarUrl: string;
+		fullname: string;
+		name: string;
+		isHf: boolean;
+		isHfAdmin: boolean;
+		isMod: boolean;
+		followerCount?: number;
+		type: "org";
+		isEnterprise: boolean;
+		isUserFollowing?: boolean;
+	}
+	| {
+		avatarUrl: string;
+		fullname: string;
+		name: string;
+		isHf: boolean;
+		isHfAdmin: boolean;
+		isMod: boolean;
+		followerCount?: number;
+		type: "user";
+		isPro: boolean;
+		_id: string;
+		isUserFollowing?: boolean;
+	};
```
Can you extract this to `ApiAuthor`?

coyotte508 reviewed (approved):
Thanks looks good!

julien-c commented:
Thanks for your contribution @quanghuynt14! 🔥
